### PR TITLE
Remove vm_init_loop

### PR DESCRIPTION
**Pull request by szilagyiadam**
For every bytecode in the above function, i created a new VM_OC group code
to be able to execute them in the vm_loop. With this update, the expectation
is a faster runtime.

JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

**Comment by szilagyiadam** [comment]
The results are the following:
```
Benchmark | Perf (sec) |
----: | ----: | 
3d-cube.js | 0.729 -> 0.734 : -0.582% | 
3d-morph.js | 0.736 -> 0.737 : -0.082% | 
3d-raytrace.js | 0.861 -> 0.862 : -0.114% | 
access-binary-trees.js | 0.448 -> 0.446 : +0.382% | 
access-fannkuch.js | 1.743 -> 1.744 : -0.063% | 
access-nbody.js | 0.978 -> 0.984 : -0.577% | 
bitops-3bit-bits-in-byte.js | 0.479 -> 0.469 : **+2.071%** | 
bitops-bits-in-byte.js | 0.647 -> 0.644 : +0.571% | 
bitops-bitwise-and.js | 0.825 -> 0.820 : +0.609% | 
bitops-nsieve-bits.js | 0.988 -> 0.984 : +0.424% | 
controlflow-recursive.js | 0.330 -> 0.311 : **+5.814%** | 
crypto-aes.js | 0.584 -> 0.588 : -0.542% | 
crypto-md5.js | 0.481 -> 0.479 : +0.600% | 
crypto-sha1.js | 0.497 -> 0.492 : +0.914% | 
date-format-tofte.js | 0.646 -> 0.645 : +0.097% | 
date-format-xparb.js | 0.489 -> 0.482 : **+1.389%** | 
math-cordic.js | 1.069 -> 1.067 : +0.184% | 
math-partial-sums.js | 0.629 -> 0.633 : -0.705% | 
math-spectral-norm.js | 0.464 -> 0.467 : -0.641% | 
string-base64.js | 0.910 -> 0.894 : **+1.799%** | 
string-fasta.js | 1.026 -> 1.024 : +0.190% | 
Geometric mean: | +0.569% | 

Binary (bytes) | master(5fdeb7c1d6) | patch(e56aceb372) | Diff |
----: | ----: | ----: | ----: |
size | 132620 | 132620 | 0 bytes |
.rodata | 11842 | 11842 | 0 bytes |
.dynstr | 212 | 212 | 0 bytes |
.rel.plt | 160 | 160 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 416 | 416 | 0 bytes |
.gnu.hash | 212 | 212 | 0 bytes |
.text | 116688 | 116416 | -272 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 96 | 96 | 0 bytes |
.plt | 268 | 268 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 32 | 32 | 0 bytes |
.bss | 1575096 | 1575096 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 180 | 180 | 0 bytes |
.gnu.version | 52 | 52 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 240 | 240 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |
```

File: jerry-core/vm/vm.c [hunk 1]
```diff
@@ -1432,14 +1194,155 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
           *stack_top_p++ = left_value;
           continue;
         }
+        case VM_OC_CREATE:
```
Comment (rerobika) [inline jerry-core/vm/vm.c]: I'd rename it to `VM_OC_CREATE_BINDING`

File: jerry-core/vm/vm.c [hunk 2]
```diff
@@ -1432,14 +1194,155 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+          {
+            lit_value = vm_construct_literal_object (frame_ctx_p,
+                                                     literal_start_p[value_index]);
+          }
+
+          if (literal_index < register_end)
+          {
+            ecma_fast_free_value (VM_GET_REGISTER (frame_ctx_p, literal_index));
+            VM_GET_REGISTER (frame_ctx_p, literal_index) = lit_value;
+            // byte_code_p = frame_ctx_p->byte_code_p;
```
Comment (rerobika) [inline jerry-core/vm/vm.c]: Please remove this line.

File: jerry-core/vm/vm.c [hunk 3]
```diff
@@ -4054,15 +3962,96 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+#endif /* ENABLED (JERRY_ES2015) */
+
+/**
+ * Resume initiaization and execution of a code block.
+ *
+ * @return ecma value
+ */
+static ecma_value_t JERRY_ATTR_NOINLINE
+vm_init_and_exec (vm_frame_ctx_t *frame_ctx_p, /**< frame context */
+                 const ecma_value_t *arg_p, /**< arguments list */
```
Comment (rerobika) [inline jerry-core/vm/vm.c]: misaligned lines.

File: jerry-core/vm/vm.c [hunk 4]
```diff
@@ -4054,15 +3962,96 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+  }
+} /* vm_execute */
+#endif /* ENABLED (JERRY_ES2015) */
+
+/**
+ * Resume initiaization and execution of a code block.
+ *
+ * @return ecma value
+ */
+static ecma_value_t JERRY_ATTR_NOINLINE
```
Comment (rerobika) [inline jerry-core/vm/vm.c]: Comment is outdated.

File: jerry-core/vm/vm.c [hunk 5]
```diff
@@ -4054,15 +3962,96 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+vm_init_and_exec (vm_frame_ctx_t *frame_ctx_p, /**< frame context */
```
Comment (rerobika) [inline jerry-core/vm/vm.c]: How about these renamings:

 - vm_init_and_exec -> vm_execute
 - vm_execute -> vm_resume_exec
?

File: jerry-core/vm/vm.h [hunk 6]
```diff
@@ -213,6 +213,11 @@ typedef enum
   VM_OC_CONTEXT_END,             /**< context end */
   VM_OC_JUMP_AND_EXIT_CONTEXT,   /**< jump and exit context */
 
+  VM_OC_CREATE_BINDING,          /**< create variables */
+  VM_OC_SET_BYTECODE_PTR,        /**< setting bytecode pointer */
+  VM_OC_VAR_FUNC_EVAL,           /**< variable and function evaluation */
+  VM_OC_INIT_LOCAL,              /**< initialize local variable */
```
Comment (rerobika) [inline jerry-core/vm/vm.h]: Imo, this group opcode is noly used in es2015 profile. Please check/guard it accroding to it.

File: jerry-core/parser/js/byte-code.h [hunk 7]
```diff
@@ -303,23 +303,23 @@
...
+              VM_OC_CREATE_BINDING) \
   CBC_OPCODE (CBC_CREATE_LOCAL, CBC_HAS_LITERAL_ARG, 0, \
-              VM_OC_NONE) \
+              VM_OC_CREATE_BINDING) \
   CBC_OPCODE (CBC_INIT_LOCAL, CBC_HAS_LITERAL_ARG | CBC_HAS_LITERAL_ARG2, 0, \
-              VM_OC_INIT_LOCALS) \
+              VM_OC_INIT_LOCAL) \
   CBC_OPCODE (CBC_CREATE_VAR_EVAL, CBC_HAS_LITERAL_ARG, 0, \
-              VM_OC_NONE) \
+              VM_OC_VAR_FUNC_EVAL) \
```
Comment (zherczeg) [inline jerry-core/parser/js/byte-code.h]: This name is strange.

File: jerry-core/vm/vm.h (hunk 6 above)
Comment (zherczeg) [reply jerry-core/vm/vm.h]: I think some of them is used in ES5.1. But it is worth to check them.

File: jerry-core/vm/vm.c [hunk 8]
```diff
@@ -1432,14 +1194,154 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
           continue;
         }
+        case VM_OC_CREATE_BINDING:
+        {
+#if !ENABLED (JERRY_ES2015)
+          if (opcode == CBC_CREATE_LET
+              || opcode == CBC_CREATE_CONST
+              || opcode == CBC_CREATE_LOCAL)
+          {
+            continue;
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: We should replace this code with an assertion here. Only CBC_CREATE_VAR is allowed.

File: jerry-core/vm/vm.c [hunk 9]
```diff
@@ -4054,15 +3961,96 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+ * Resume execution of a code block.
  *
  * @return ecma value
  */
-static void JERRY_ATTR_NOINLINE
-vm_init_exec (vm_frame_ctx_t *frame_ctx_p, /**< frame context */
-              const ecma_value_t *arg_p, /**< arguments list */
-              ecma_length_t arg_list_len) /**< length of arguments list */
+ecma_value_t JERRY_ATTR_NOINLINE
+vm_resume_exec (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: The code below looks like a duplication. That should be really avoided.

File: jerry-core/vm/vm.h [hunk 10]
```diff
@@ -414,11 +417,12 @@ ecma_value_t vm_run_module (const ecma_compiled_code_t *bytecode_p, ecma_object_
 
 ecma_value_t vm_run (const ecma_compiled_code_t *bytecode_header_p, ecma_value_t this_binding_value,
                      ecma_object_t *lex_env_p, const ecma_value_t *arg_list_p, ecma_length_t arg_list_len);
-ecma_value_t vm_execute (vm_frame_ctx_t *frame_ctx_p);
 
 bool vm_is_strict_mode (void);
 bool vm_is_direct_eval_form_call (void);
 
+ecma_value_t vm_execute (vm_frame_ctx_t *frame_ctx_p);
```
Comment (zherczeg) [inline jerry-core/vm/vm.h]: Why this declaration is moved down here?

File: jerry-core/vm/vm.c [hunk 11]
```diff
@@ -4054,11 +3956,6 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
 #undef READ_LITERAL
 #undef READ_LITERAL_INDEX
 
-/**
- * Initialize code block execution
- *
- * @return ecma value
- */
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: Why this comment is removed?

File: jerry-core/vm/vm.c [hunk 12]
```diff
@@ -4148,12 +4050,10 @@ vm_init_exec (vm_frame_ctx_t *frame_ctx_p, /**< frame context */
 
   JERRY_CONTEXT (status_flags) &= (uint32_t) ~ECMA_STATUS_DIRECT_EVAL;
   JERRY_CONTEXT (vm_top_context_p) = frame_ctx_p;
-
-  vm_init_loop (frame_ctx_p);
 } /* vm_init_exec */
 
 /**
- * Resume execution of a code block.
+ * Initiaization and execution of a code block.
```
Comment (rerobika) [inline jerry-core/vm/vm.c]: Please do not modify this comment

**Pull request merged**